File: sparkle-studio/backend/schemas/git.py
"""
Pydantic schemas for Git operations.
"""
from typing import Optional
from datetime import datetime
from pydantic import BaseModel


class GitBranch(BaseModel):
    """Git branch information."""
    name: str
    is_current: bool = False
    commit_sha: Optional[str] = None
    last_commit_message: Optional[str] = None
    last_commit_date: Optional[datetime] = None


class GitStatus(BaseModel):
    """Git repository status."""
    current_branch: str
    is_clean: bool
    modified_files: list[str] = []
    untracked_files: list[str] = []
    staged_files: list[str] = []
    ahead: int = 0  # Commits ahead of remote
    behind: int = 0  # Commits behind remote
    remote_url: Optional[str] = None


class GitCommitRequest(BaseModel):
    """Request to commit changes."""
    message: str
    files: Optional[list[str]] = None  # If None, commit all changes
    push: bool = False  # Auto-push after commit


class GitCommitResponse(BaseModel):
    """Response from commit operation."""
    commit_sha: str
    message: str
    author: str
    timestamp: datetime
    files_changed: list[str] = []
    pushed: bool = False


class GitPullRequest(BaseModel):
    """Request to pull from remote."""
    branch: Optional[str] = None  # If None, use current branch
    rebase: bool = False


class GitPullResponse(BaseModel):
    """Response from pull operation."""
    success: bool
    branch: str
    commits_pulled: int
    files_changed: list[str] = []
    message: str


class GitPRRequest(BaseModel):
    """Request to create a pull request."""
    title: str
    description: str
    source_branch: str
    target_branch: str = "main"
    draft: bool = False


class GitPRResponse(BaseModel):
    """Response from PR creation."""
    pr_number: int
    pr_url: str
    title: str
    source_branch: str
    target_branch: str

File: sparkle-studio/frontend/src/types/git.ts
/**
 * Git type definitions matching backend schemas.
 */

export interface GitBranch {
  name: string;
  is_current: boolean;
  commit_sha?: string;
  last_commit_message?: string;
  last_commit_date?: string;
}

export interface GitStatus {
  current_branch: string;
  is_clean: boolean;
  modified_files: string[];
  untracked_files: string[];
  staged_files: string[];
  ahead: number;
  behind: number;
  remote_url?: string;
}

export interface GitCommitRequest {
  message: string;
  files?: string[];
  push: boolean;
}

export interface GitCommitResponse {
  commit_sha: string;
  message: string;
  author: string;
  timestamp: string;
  files_changed: string[];
  pushed: boolean;
}

export interface GitPRRequest {
  title: string;
  description: string;
  source_branch: string;
  target_branch: string;
  draft: boolean;
}

export interface GitPRResponse {
  pr_number: number;
  pr_url: string;
  title: string;
  source_branch: string;
  target_branch: string;
}
